homeassistant/components/insteon/__init__.py

```diff
@@ -92,57 +97,57 @@ async def close_insteon_connection(*args):
     await async_close()
 
 
+async def async_import_config(hass, conf):
+    """Set up all of the config imported from yaml."""
+    data, options = convert_yaml_to_config_flow(conf)
+    # Create a config entry with the connection data
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": SOURCE_IMPORT}, data=data
+    )
+    # If this is the first time we ran, update the config options
+    if result.get("result") and options:
+        entry = result["result"]
+        hass.config_entries.async_update_entry(
+            entry=entry, options=options,
+        )
+    return result
```

Comment:
```suggestion
```

I don't think anything is consuming the return

Reply:
Turns out this is important. It returns processing back to the config import process. Without returning the result, the import process does not know about the process being completed.

Reply:
Can you point me to where it is being checked?

Reply:
Since `async_import_config` isn't part of the config_flow.py, it really should be tested in `test_init.py` and the config flow should test the import without calling `async_import_config` with something like (borrowed from test_step_import in `nexia`):

```
    result = await hass.config_entries.flow.async_init(
        DOMAIN,
        context={"source": config_entries.SOURCE_IMPORT},
        data={CONF_USERNAME: "username", CONF_PASSWORD: "password"},
    )
```

To test `async_import_config` in `test_init.py`, we should go though the standard interface and call `async_setup_component` to verify the import happens.  You could patch `async_setup_entry` to make sure it doesn't actually setup and then check the config entries to make sure the data and options are populated.

Reply:
This code is being called in tests otherwise these tests would fail:
https://github.com/teharris1/home-assistant/blob/config_flow/tests/components/insteon/test_config_flow.py#L275
https://github.com/teharris1/home-assistant/blob/config_flow/tests/components/insteon/test_config_flow.py#L292
and
https://github.com/teharris1/home-assistant/blob/config_flow/tests/components/insteon/test_config_flow.py#L312

Reply:
Oh, actually, the test does use `async_setup_component` directly, so I think I am doing that already.

Reply:
> Oh, actually, the test does use `async_setup_component` directly, so I think I am doing that already.

It is used for `persistent_notification`, but its not actually setting up this integration (`insteon`).
`await setup.async_setup_component(hass, "persistent_notification", {})`

The key issue here is that we are testing `async_import_config` directly instead of going though the standard interfaces so that doesn't give us a full end to end test.

Reply:
I see what you are saying but I still feel this is correct. For example, if I look at the Hue component, inside `async_setup` it makes a call to 
```
        hass.async_create_task(
            hass.config_entries.flow.async_init(
                DOMAIN,
                context={"source": config_entries.SOURCE_IMPORT},
                data={"host": bridge_conf[CONF_HOST]},
            )
        )
``` 
which is exactly what you would expect in order to import the configuration into a config flow. So a test would need to call `hass.config_entries.flow.async_init` as you describe above.   What I am doing inside `async_setup` is
```
    hass.async_create_task(async_import_config(hass, conf))
```
Where `conf` is the configuration directly from yaml. Inside `async_import_config` I call `hass.config_entries.flow.async_init`. I needed to do this in order to import options at the same time as the main configuration because a config flow import step has no ability to import options. So I had to capture the return from the import step after the config is setup. I believe this is more complicated than it should be but I saw no way around it given what appears to be limitations of the config flow import process. So, my test calls `async_import_config` here:
https://github.com/teharris1/home-assistant/blob/0438790f07e4ffc39240fab9948da2ae2e7ed0cc/tests/components/insteon/test_config_flow.py#L260
So I believe this is testing the exact flow that HA would run at startup.  Perhaps some code comments would have helped make the flow easier to understand?

Reply:
Rereading my comment I feel I could be more clear. Inside `async_setup`, I am calling `async_import_config`. This is just a wrapper for `hass.config_entries.flow.async_init` in order to do some pre and post-processing that is not possible inside config flow. So where a normal test would be to call `hass.config_entries.flow.async_init` the right test for my code is to call `async_import_config`. Does that help?

Reply:
The concern I'm trying to articulate is that the config flow tests (`test_config_flow.py`) are being mixed in with the init tests (`test_init.py`) because `test_init.py` doesn't exist.

The `async_import_config` code isn't a normal pattern used in the code base.  I see that you are using it to split up the options and data from the yaml so it makes sense.  Normally we would want to call something like this in `test_step_import_....` in `config_flow.py` with the mock data already in the import format:

```
    result = await hass.config_entries.flow.async_init(
        DOMAIN, context={"source": SOURCE_IMPORT}, data=data
    )
```

The way it is now, we are testing how `async_import_config` behaves in the `test_config_flow.py` instead of `test_init.py`.  Assuming `DOMAIN = "insteon"` Ideally we test this with something like this in `test_init.py`:

```
with patch("homeassistant.components.insteon.async_setup_entry", return_value=True):
    assert await async_setup_component(hass, DOMAIN, <CONFIG_FROM_YAML_HERE>)
    await hass.async_block_till_done()

```
Then check to make sure setup did what you expected.

On a side note, there is talk about combining `data` and `options` (in other discussions). I think something like `async_import_config` hopefully won't be needed anymore.

Reply:
After writing that all out, we are still going to end up with a compromise either way because it is not possible to import `data` and `options` at the same time.

If another reviewer has suggestions on how to improve this, it can be done in a followup PR.

Reply:
As already said, this is not a pattern we use currently. Instead we should just check if there are options set when setting up the config entry and set defaults at that time. The config can be stored in hass.data if needed.